lib/src/main/kotlin/tsl/trees/AVLTree.kt

```diff
@@ -0,0 +1,159 @@
+        return node.height
+    }
+
+    private fun getBalanceFactor(node: AVLNode<K, V>): Int =
+        getHeight(node.rightChild) - getHeight(node.leftChild)
+
+    private fun getHeight(node: AVLNode<K, V>?): Int = node?.height ?: -1
+
+    public override fun delete(key: K): V? {
+        val deletedValue: V =
+            search(key) ?: return null // if key isn't in the tree, there's nothing to delete
+
+        deleteAndBalanceRecursively(root, key)
+
+        return deletedValue
+    }
+
+    private fun deleteAndBalanceRecursively(
+        currNode: AVLNode<K, V>?,
+        keyToDelete: K
+    ): AVLNode<K, V>? {
+
+        when {
+            currNode == null -> return null // node to be deleted was not found
+            keyToDelete < currNode.key -> // node to be deleted is in the left subtree
+            currNode.leftChild = deleteAndBalanceRecursively(currNode.leftChild, keyToDelete)
+            keyToDelete > currNode.key -> // node to be deleted is in the right subtree
+            currNode.rightChild = deleteAndBalanceRecursively(currNode.rightChild, keyToDelete)
+            else -> {
+                if (currNode.leftChild == null || currNode.rightChild == null) {
+                    return currNode.leftChild ?: currNode.rightChild
+                } else {
+                    val successor = getMinNodeRecursively(currNode.rightChild)
+                    if (successor != null) {
+                        // copy its successor
+                        currNode.key = successor.key
+                        currNode.value = successor.value
+
+                        // delete original successor node from tree
+                        val newSubtree =
+                            deleteAndBalanceRecursively(currNode.rightChild, successor.key)
+                        if (newSubtree != null) currNode.rightChild = newSubtree
+                    }
+                }
+            }
+        }
+
+        var balancedNode: AVLNode<K, V> = currNode
+
+        if (getBalanceFactor(currNode) < -1) { // if left subtree got bigger
+            currNode.leftChild?.let {
+                if (getBalanceFactor(it) > 0) { // if inserted node has greater key
+                    currNode.leftChild = rotateLeft(it) // perform left rotation first
+                }
+                balancedNode = rotateRight(currNode) // anyway, perform left rotation
+            }
+        } else if (getBalanceFactor(currNode) > 1) { // if right subtree got bigger
+            currNode.rightChild?.let {
+                if (getBalanceFactor(it) < 0) { // if inserted node has greater key
+                    currNode.rightChild = rotateRight(it) // perform right rotation first
+                }
+                balancedNode = rotateLeft(currNode) // anyway, perform left rotation
+            }
+        }
+
+        updateHeight(balancedNode.leftChild)
+        updateHeight(balancedNode.rightChild)
+        updateHeight(balancedNode)

```

Comment:
```suggestion
        updateHeightWithChildren(balancedNode)
```
субъективно, но можно было в 1 функцию объединить. плюс, встречается в insert'е ранее